src/content/services/monetizationLinkManager.ts

```diff
@@ -498,56 +498,40 @@ export class MonetizationLinkManager extends EventEmitter {
     void this.sendStartMonetization(startMonetizationPayload);
   }
 
-  private async onAddedNode(
-    node: Node,
-  ): Promise<StartMonetizationPayload | null> {
-    if (node instanceof HTMLElement) {
-      this.dispatchOnMonetizationAttrChangedEvent(node);
-    }
+  private async onAddedLink(
+    link: HTMLLinkElement,
+  ): Promise<StartMonetizationPayloadEntry | null> {
+    if (this.monetizationLinks.has(link)) {
+      const details = this.monetizationLinks.get(link);
+      if (!details) {
+        throw new Error(
+          `Could not find details promise for monetization node ${link.outerHTML.slice(0, 200)}`,
+        );
+      }
 
-    if (node instanceof HTMLLinkElement) {
-      const payloadEntry = await this.onAddedLink(node);
-      return payloadEntry ? [payloadEntry] : null;
-    } else if (node instanceof HTMLElement) {
-      const linkElements = this.getMonetizationLinkTags(node);
-      return await Promise.all(
-        linkElements.map((linkElem) => this.onAddedLink(linkElem)),
-      ).then((res) => res.filter(isNotNull));
+      return details;
     }
-    return null;
-  }
 
-  private onRemovedNode(node: Node): StopMonetizationPayload | null {
-    if (node instanceof HTMLLinkElement) {
-      return [this.onRemovedLink(node)];
-    } else if (node instanceof HTMLElement) {
-      const linkElements = this.getMonetizationLinkTags(node).filter((el) =>
-        this.monetizationLinks.has(el),
-      );
-      return linkElements.map((linkElem) => this.onRemovedLink(linkElem));
+    const promise = this.checkLink(link);
+    this.monetizationLinks.set(link, promise);
+
+    const response = await promise;
+    if (response) {
+      this.observeLinkAttrs(link);
     }
 
-    return null;
+    return promise;
```

Comment:
I think we're making the wrong part async/promise-cache. The real async call in call stack that we want to _cache_ is `validateLink`, which does the message passing bit. Let's try caching there instead?